tls/s2n_ktls.c

```diff
@@ -82,38 +104,147 @@ S2N_RESULT s2n_ktls_retrieve_file_descriptor(struct s2n_connection *conn, s2n_kt
     } else if (ktls_mode == S2N_KTLS_MODE_SEND) {
         RESULT_GUARD_POSIX(s2n_connection_get_write_fd(conn, fd));
     }
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_ktls_get_io_mode(s2n_ktls_mode ktls_mode, int *tls_tx_rx_mode)
+{
+    RESULT_ENSURE_REF(tls_tx_rx_mode);
+
+    if (ktls_mode == S2N_KTLS_MODE_SEND) {
+        *tls_tx_rx_mode = S2N_TLS_TX;
+    } else {
+        *tls_tx_rx_mode = S2N_TLS_RX;
+    }
+    return S2N_RESULT_OK;
+}
 
+#if defined(S2N_KTLS_SUPPORTED)
+/* If server is sending or client is receiving then use server key material */
+static S2N_RESULT s2n_use_server_key_material(struct s2n_connection *conn, s2n_ktls_mode ktls_mode, bool *use_server_km)
+{
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(use_server_km);
+    bool server_sending = (conn->mode == S2N_SERVER && ktls_mode == S2N_KTLS_MODE_SEND);
+    bool client_receiving = (conn->mode == S2N_CLIENT && ktls_mode == S2N_KTLS_MODE_RECV);
+    *use_server_km = server_sending || client_receiving;
     return S2N_RESULT_OK;
 }
 
-static S2N_RESULT s2n_ktls_configure_socket(struct s2n_connection *conn, s2n_ktls_mode ktls_mode)
+S2N_RESULT s2n_ktls_init_aes128_gcm_crypto_info(struct s2n_connection *conn, s2n_ktls_mode ktls_mode,
+        struct s2n_key_material *key_material, struct tls12_crypto_info_aes_gcm_128 *crypto_info)
 {
     RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(conn->client);
+    RESULT_ENSURE_REF(conn->server);
+    RESULT_ENSURE_REF(key_material);
+    RESULT_ENSURE_REF(crypto_info);
+    RESULT_ENSURE_REF(conn->secure);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite);
+    RESULT_ENSURE_REF(conn->secure->cipher_suite->record_alg);
+    const struct s2n_cipher *cipher = conn->secure->cipher_suite->record_alg->cipher;
+    RESULT_ENSURE_REF(cipher);
+
+    RESULT_ENSURE(cipher == &s2n_aes128_gcm, S2N_ERR_KTLS_UNSUPPORTED_CONN);
+    RESULT_ENSURE(conn->actual_protocol_version == S2N_TLS12, S2N_ERR_KTLS_UNSUPPORTED_CONN);
+    crypto_info->info.cipher_type = TLS_CIPHER_AES_GCM_128;
+    crypto_info->info.version = TLS_1_2_VERSION;
+
+    /* set values based on mode of operation */
+    struct s2n_blob *key = NULL;
+    struct s2n_blob implicit_iv = { 0 };
+    struct s2n_blob sequence_number = { 0 };
+
+    bool use_server_km = false;
+    RESULT_GUARD(s2n_use_server_key_material(conn, ktls_mode, &use_server_km));
+    if (use_server_km) {
```

Comment:
Looking at this again, the separate method probably isn't required. You could just go:
```suggestion
bool server_sending = (conn->mode == S2N_SERVER && ktls_mode == S2N_KTLS_MODE_SEND);
bool client_receiving = (conn->mode == S2N_CLIENT && ktls_mode == S2N_KTLS_MODE_RECV);
if (server_sending || client_receiving) {
```

Reply:
yea its much neater once separated